### PR TITLE
Design and implementation of scenario tree

### LINKED ISSUE #1: scenario tree implementation
### Description
- [x] create scenario tree class
- [ ] define different ways of generating tree

### Functions
- [ ] DO FIRST -> decide how these will be returned (lists, arrays, ... ?)
- [ ] return number of nodes
- [ ] return node ancestor
- [ ] return node children
- [ ] return number of node children
- [ ] return nodes at time stage
- [ ] return number of nodes at time stage

## Tests

How to test it

### PR DESCRIPTION
**Pull request by alphaville**
## About 

Design and implementation of scenario tree. @Gczmy and @smokinmirror you are invited to participate. I have written a few unit tests (see [TDD](https://en.wikipedia.org/wiki/Test-driven_development)).

## Issues

- Closes #1 

## TODO

Before we merge this, make sure that:

- [x] all public methods are documented
- [x] we have suffient tests (we need a test of `set_data_at_node` and `get_data_at_node`)
- [x] all tests pass 
- [x] we have merged #6
- [x] we have merged #7

### CODE REVIEW

**Comment by alphaville** [comment]
@Gczmy This is almost ready now. We just need to (i) implement at least one more test, (ii) document all methods. @Gczmy and @smokinmirror I have requested your review on this. Let us try to get this done soon, merge this branch and close #1.

**Review by Gczmy — APPROVED** [review]
All test pass.

**Review by Gczmy — APPROVED** [review]
All tests pass.

File: ci/script.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    # --- create virtualenv
+    virtualenv -p python3.8 venv
+
+    # --- activate venv
+    source venv/bin/activate
+
+    # --- upgrade pip within venv
+    pip install --upgrade pip
+
+    # --- install opengen
```
Comment (ruairimoran) [inline ci/script.sh]: opengen ?

File: raocp/core/scenario_tree.py [hunk 2]
```diff
@@ -1,5 +1,326 @@
...
+
+
 class ScenarioTree:
+    """
+    Scenario tree
+    """
+
+    def __init__(self, stages, ancestors, probability, w_values=None):
+        """
+        :param stages:
```
Comment (ruairimoran) [inline raocp/core/scenario_tree.py]: bulk out docs

Comment (ruairimoran) [inline raocp/core/scenario_tree.py]: bulk out docs

File: raocp/core/scenario_tree.py [hunk 3]
```diff
@@ -1,5 +1,326 @@
...
+        :return: stored data
+        """
+        return self.__data[node_idx]
+
+    def set_data_at_node(self, node_idx, data_dict: dict):
+        """
+        :param node_idx: node index
+        :param data_dict: a dictionary with the data to be stored at the above node
+
+        Returns `None` if no data are stored at the given node
```
Comment (ruairimoran) [inline raocp/core/scenario_tree.py]: move line to `get_data_at_node` doc above

File: raocp/core/scenario_tree.py [hunk 4]
```diff
@@ -1,5 +1,326 @@
...
+        """
+        self.__data[node_idx] = data_dict
+
+    def num_nodes(self):
+        """
+        :return: total number of nodes of the tree
+        """
+        return len(self.__ancestors)
```
Comment (ruairimoran) [inline raocp/core/scenario_tree.py:77]: double check

Comment (ruairimoran) [reply raocp/core/scenario_tree.py:77]: may need +1 for root node

File: raocp/core/scenario_tree.py [hunk 5]
```diff
@@ -1,5 +1,326 @@
...
+    """
+    Factory class to construct scenario trees from stopped Markov chains
+    """
+
+    def __init__(self, transition_prob, initial_distribution, num_stages, stopping_time=None):
+        """
+        :param transition_prob: transition matrix of the Markov chain
+        :param initial_distribution: initial distribution of `w`
+        :param num_stages: total number of stages or horizon of the scenario tree
+        :param stopping_time: stopping time, which much be no larger than the number of stages [default: None]
```
Comment (ruairimoran) [inline raocp/core/scenario_tree.py]: much -> must

File: tests/test_scenario_tree.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
+import unittest
```
Comment (ruairimoran) [inline tests/test_scenario_tree.py:1]: failure test for when `tau > N`

**Review by ruairimoran — CHANGES_REQUESTED** [review]
@Gczmy couple of tweaks and a unit test you should be able to implement

File: ci/script.sh (hunk 1 above)
Comment (ruairimoran) [reply ci/script.sh]: changed to `raocp-toolbox`

File: raocp/core/scenario_tree.py (hunk 2 above)
Comment (ruairimoran) [reply raocp/core/scenario_tree.py]: done

File: raocp/core/scenario_tree.py (hunk 3 above)
Comment (ruairimoran) [reply raocp/core/scenario_tree.py]: done

File: raocp/core/scenario_tree.py (hunk 4 above)
Comment (ruairimoran) [reply raocp/core/scenario_tree.py:77]: don't need to. root node is given ancestor -1

File: raocp/core/scenario_tree.py (hunk 5 above)
Comment (ruairimoran) [reply raocp/core/scenario_tree.py]: done

File: tests/test_scenario_tree.py (hunk 6 above)
Comment (ruairimoran) [reply tests/test_scenario_tree.py:1]: done

**Review by ruairimoran — APPROVED** [review]
on to the next thing

**Pull request merged**